daemon/logger/loggerutils/logfile.go

```diff
@@ -587,27 +784,58 @@ func tailFiles(files []SizeReaderAt, watcher *logger.LogWatcher, dec Decoder, ge
 		}
 	}()
 
-	readers := make([]io.Reader, 0, len(files))
+	readers, err := getTailFiles(ctx, files, nLines, getTailReader)
+	if err != nil {
+		watcher.Err <- err
+		return false
+	}
 
-	if nLines > 0 {
-		for i := len(files) - 1; i >= 0 && nLines > 0; i-- {
-			tail, n, err := getTailReader(ctx, files[i], nLines)
-			if err != nil {
-				watcher.Err <- errors.Wrap(err, "error finding file position to start log tailing")
-				return false
+	var idx int
+	defer func() {
+		// Make sure all are released if there is an early return.
+		if !cont {
+			for _, r := range readers[idx:] {
+				if err := r.Close(); err != nil {
+					log.G(ctx).WithError(err).Debug("Error closing log reader")
+				}
 			}
-			nLines -= n
-			readers = append([]io.Reader{tail}, readers...)
 		}
-	} else {
-		for _, r := range files {
-			readers = append(readers, r)
+	}()
+
+	// Wrap the decoder so that decode errors return an EOF
+	wrapDecoder := func(dec Decoder) decoderFunc {
+		return func() (*logger.Message, error) {
+			msg, err := dec.Decode()
+			if err != nil && !errors.Is(err, io.EOF) {
+				err = fmt.Errorf("%w: %w", io.EOF, err)
```

Comment:
Decoders don't have control over underlying I/O though so I'm not really sure they can be cancellable.
What we can do is start a goroutine waiting for the parent to cancel and then close the file, which should trigger the decode to return EOF/UnexpectedEOF.

Reply:
A robust decoder for a self-synchronizing log format could test if the context is done between the `Read()`s to scan for the next record delimiter and exit the loop early. That said, I like your idea better as it would work unmodified with any decoder implementation that handles `fs.ErrClosed` by returning.